internal/sourcemap/metadata.go

```diff
@@ -0,0 +1,235 @@
+
+func NewMetadataCachingFetcher(
+	c *elasticsearch.Client,
+	backend Fetcher,
+	index string,
+	invalidationChan chan<- Identifier,
+) *MetadataCachingFetcher {
+	return &MetadataCachingFetcher{
+		esClient:         c,
+		index:            index,
+		set:              make(map[Identifier]string),
+		backend:          backend,
+		logger:           logp.NewLogger(logs.Sourcemap),
+		invalidationChan: invalidationChan,
+	}
+}
+
+func (s *MetadataCachingFetcher) Fetch(ctx context.Context, name, version, path string) (*sourcemap.Consumer, error) {
+	key := Identifier{
+		name:    name,
+		version: version,
+		path:    path,
+	}
+
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if _, found := s.set[key]; found {
+		// Only fetch from ES if the sourcemap id exists
+		return s.backend.Fetch(ctx, name, version, path)
+	}
+
+	// Try again
+	key.path = maybeParseURLPath(path)
+
+	if _, found := s.set[key]; found {
+		// Only fetch from ES if the sourcemap id exists
+		return s.backend.Fetch(ctx, name, version, path)
+	}
+
+	return nil, nil
+}
+
+// maybeParseURLPath attempts to parse s as a URL, returning its path component
+// if successful. If s cannot be parsed as a URL, s is returned.
+func maybeParseURLPath(s string) string {
+	url, err := url.Parse(s)
+	if err != nil {
+		return s
+	}
+	return url.Path
+}
+
+func (s *MetadataCachingFetcher) update(updates map[Identifier]string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for id, contentHash := range s.set {
+		if updatedHash, ok := updates[id]; ok {
+			// already in the cache, remove from the updates.
+			delete(updates, id)
+
+			// content hash changed, invalidate the sourcemap cache
+			if contentHash != updatedHash {
+				s.invalidationChan <- id
+			}
+		} else {
+			// the sourcemap no longer exists in ES.
+			// invalidate the sourcemap cache.
+			s.invalidationChan <- id
+
+			// remove from metadata cache
+			delete(s.set, id)
+		}
+	}
+	// add new sourcemaps to the metadata cache.
+	for id, contentHash := range updates {
+		s.set[id] = contentHash
+	}
+}
+
+func (s *MetadataCachingFetcher) StartBackgroundSync() {
+	go func() {
+		// First run, populate cache
+		ctx, cleanup := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cleanup()
+
+		if err := s.sync(ctx); err != nil {
+			s.logger.Error("failed to fetch sourcemaps metadata: %v", err)
+		}
+	}()
+
+	go func() {
+		// TODO make this a config option ?
+		t := time.NewTicker(30 * time.Second)
+		defer t.Stop()
+
+		for range t.C {
+			ctx, cleanup := context.WithTimeout(context.Background(), 10*time.Second)
+
+			if err := s.sync(ctx); err != nil {
+				s.logger.Error("failed to sync sourcemaps metadata: %v", err)
+			}
+
+			cleanup()
+		}
+	}()
+}
+
+func (s *MetadataCachingFetcher) sync(ctx context.Context) error {
+	resp, err := s.runSearchQuery(ctx)
+	if err != nil {
+		return errors.Wrap(err, errMsgESFailure)
+	}
+	defer resp.Body.Close()
+
+	// handle error response
+	if resp.StatusCode >= http.StatusMultipleChoices {
```

Comment:
Why is it >=300 instead of >=400?

Reply:
That's what the current elasticsearch fetcher is using: https://github.com/elastic/apm-server/blob/da18d54ecdb3f442d973007ebe2ebdd33692a143/internal/sourcemap/elasticsearch.go#L84